onnxruntime/core/providers/dml/DmlExecutionProvider/src/TensorDesc.cpp

```diff
@@ -315,3 +315,60 @@ void TensorDesc::SetDimensionCount(uint32_t newDimensionCount, TensorAxis alignm
     }
     m_bufferTensorDesc.DimensionCount = newDimensionCount;
 }
+
+// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
+void TensorDesc::PermuteDimensions(gsl::span<const uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    EnsureMinimumDimensionCount(static_cast<uint32_t>(dimensionMapping.size()), alignment);
+    EnsureStridesExist(static_cast<uint32_t>(dimensionMapping.size()), alignment);
+    PermuteArray(dimensionMapping, alignment);
+}
+
+// Shuffle m_sizes and m_strides according to the indexes pointed by dimensionMapping
+void TensorDesc::PermuteArray(gsl::span<const uint32_t> dimensionMapping, TensorAxis alignment)
+{
+    std::vector<uint32_t> tempSizes(dimensionMapping.size());
+    std::vector<uint32_t> tempStrides(dimensionMapping.size());
+    // Right alignment values are shifted to the end
```

Comment:
>Right alignment values are shifted to the end [](http://example.com/codeflow?start=7&length=45)

Dimensions should be aligned to the right edge of the new count, not the right edge of the maximum possible size.

So if your original dimensions are:

`originalSize = {2,3}`

And you want a new size of at least 4, left aligned:

`newSize = {2,3,1,1}`

For right aligned:

`newSize = {1,1,2,3}`

But in both cases, the last 4 values in `m_sizes` remains unused 0's.

`m_sizes = {2,3,1,1,0,0,0,0} // left aligned`
`m_sizes = {1,1,2,3,0,0,0,0} // right aligned`